frontend/src/libs/components/note/note.tsx

```diff
@@ -0,0 +1,175 @@
+  useAppDispatch,
+  useAppForm,
+  useAppSelector,
+  useCallback,
+  useEffect,
+  useFormController,
+  useParams,
+  useRef,
+} from '#libs/hooks/hooks.js';
+import { type UserAuthResponseDto } from '#packages/users/users.js';
+import { actions as journalActions } from '#slices/journal/journal.js';
+
+import { NOTE_TIMEOUT } from './libs/constants.js';
+import { type NoteContent } from './libs/types.js';
+import styles from './styles.module.scss';
+
+type Properties = {
+  className: string;
+};
+
+const Note: React.FC<Properties> = ({ className }) => {
+  const { userId, selectedJournalEntry } = useAppSelector(
+    ({ auth, journal }) => {
+      return {
+        userId: (auth.authenticatedUser as UserAuthResponseDto).id,
+        selectedJournalEntry: journal.selectedJournalEntry,
+      };
+    },
+  );
+
+  const { isDirty, control } = useAppForm({
+    defaultValues: {
+      title: selectedJournalEntry?.title,
+      text: selectedJournalEntry?.text,
+    },
+    mode: 'onChange',
+  });
+
+  const { field: titleField } = useFormController({ name: 'title', control });
+  const { field: textField } = useFormController({ name: 'text', control });
+  const { value: titleValue, onChange: onTitleChange } = titleField;
+  const { value: textValue, onChange: onTextChange } = textField;
+
+  const dispatch = useAppDispatch();
+  const { id } = useParams();
+
+  const titleReference = useRef<HTMLDivElement | null>(null);
+  const textReference = useRef<HTMLDivElement | null>(null);
+  const cursorPosition = useRef<number | null>(null);
+
+  const handleSaveNote = useCallback(
+    (data: NoteContent) => {
+      if (data.title && data.text && id) {
+        void dispatch(
+          journalActions.updateJournalEntry({
+            id: Number(id),
+            body: {
+              title: sanitizeInput(data.title),
+              text: sanitizeInput(data.text),
+            },
+          }),
+        );
+      }
+    },
+    [dispatch, id],
+  );
+
+  const handleNoteChange = useCallback(
+    (
+      elementReference: React.MutableRefObject<HTMLDivElement | null>,
+      onChange: (value: string) => void,
+    ): void => {
+      if (elementReference.current) {
+        const newValue = elementReference.current.textContent ?? '';
+        onChange(newValue);
+        changeCursorPosition(cursorPosition);
+      }
+    },
+    [],
+  );
+
+  const handleTitleChange = useCallback(() => {
+    handleNoteChange(titleReference, onTitleChange);
+  }, [handleNoteChange, onTitleChange]);
+
+  const handleTextChange = useCallback(() => {
+    handleNoteChange(textReference, onTextChange);
+  }, [handleNoteChange, onTextChange]);
+
+  useEffect(() => {
+    void dispatch(journalActions.getAllJournalEntries()).then(() => {
+      if (id) {
+        void dispatch(journalActions.setSelectedJournalEntry(Number(id)));
+      }
+    });
+  }, [dispatch, id]);
+
+  useEffect(() => {
+    const handleSaveNoteWithDebounce = debounce((data: NoteContent) => {
```

Comment:
looks like this won't work, because we will create new debounce function on every change. but we need to call this function only after the last change.

For example: user types something, if less than 2 seconds passed between new changes, nothing saves. Only after they stop typing (more than 2 seconds passed), request is sent

I'm not sure if we don't lose reference in debounce, looks like we should wrap it into custom hook 🤔

Reply:
But it works as expected now. It calls the debounced function only in 2 sec when the user stops typing. When I tried to have this function outside with useCallback I had a warning that I needed to put this function without callback as its dependencies are unknown. Also, we have a similar example of debouncing usage in the search component.  You can see it in this screenshot.  So everything works as expected now.
![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/102380564/c5ff8dc8-5e34-4ea1-8718-7b2f354d5d11)

Reply:
but how can this work if debounce() always returns a new function?

Reply:
You can see how it works now here)
https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/102380564/c4146a19-596b-4f9c-b2c9-d2e3e9d3af6f

Reply:
Side effects may work sometimes but it is not a best practise

Reply:
Let's put callback into debounce without condition

Reply:
@v0ldemar01  Can you, please clarify what you mean? If like in 1 screenshot, so it sends request every 2 sec, so it doesn't work as expected. If in 2 variant I have a warning here:
![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/102380564/342d2188-1b31-488d-a984-ed498bef1250)

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/102380564/76c8d234-77a7-413d-8915-97672a9a1b60)

Reply:
You are receiving the error, bcos you missed to pass handleSaveNote as deps to handleSaveNoteWithDebounce

Reply:
if I make it like this, I don't have an error, but I can't use handleSaveNoteWithDebounce.clear() anymore. So how could I handle it?
![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/102380564/0383f3da-dec1-45b2-9258-25904fffb942)